File: backend/app/schemas/employee.py
# app/schemas/employee.py
from pydantic import BaseModel, EmailStr, Field, validator, ConfigDict
from typing import Optional
from datetime import date, datetime

# Create a base Department schema
class DepartmentBase(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)

class EmployeeBase(BaseModel):
    employee_id: str = Field(..., min_length=1, max_length=50)
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr
    department_id: Optional[int] = None
    position: Optional[str] = Field(None, max_length=100)
    hire_date: Optional[date] = None
    model_config = ConfigDict(from_attributes=True)

class EmployeeCreate(EmployeeBase):
    @validator('employee_id')
    def employee_id_alphanumeric(cls, v):
        if not v.replace('_', '').replace('-', '').isalnum():
            raise ValueError('Employee ID must be alphanumeric')
        return v

class EmployeeUpdate(BaseModel):
    first_name: Optional[str] = Field(None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(None, min_length=1, max_length=100)
    email: Optional[EmailStr] = None
    department_id: Optional[int] = None
    position: Optional[str] = Field(None, max_length=100)
    hire_date: Optional[date] = None
    is_active: Optional[bool] = None
    model_config = ConfigDict(from_attributes=True)

# Updated Employee response model with department
class Employee(EmployeeBase):
    id: int
    is_active: bool
    department: Optional[DepartmentBase] = None  # Add this line
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

class EmployeeList(BaseModel):
    employees: list[Employee]
    total: int
    skip: int
    limit: int

File: frontend/src/types/employee.ts
// employee.ts
import { type Department } from "./department";

export interface Employee {
    id: number;
    employee_id: string;
    first_name: string;
    last_name: string;
    email: string;
    department_id?: number | null;
    department?: Department;
    position?: string;
    hire_date?: string;
    is_active?: boolean;
    created_at?: string;
    updated_at?: string;
}

export interface EmployeeFormData {
    employee_id: string;
    first_name: string;
    last_name: string;
    email: string;
    department_id?: number | null;
    position: string;
    hire_date: string;
}

export interface EmployeeListResponse {
    employees: Employee[];
    total: number;
    skip: number;
    limit: number;
}
